7zip/Makefile

```diff
@@ -1,51 +1,97 @@
+SRCDIR := $(abspath $(patsubst %/,%,$(dir $(abspath $(lastword $(MAKEFILE_LIST))))))
+
 CC ?= gcc
 CXX ?= g++
 AR ?= ar
 
 SRC_CFLAGS = -W -Wall -Wextra
-SRC_CXXFLAGS = -W -Wall -Wextra -std=c++11 -ICPP
+SRC_CXXFLAGS = -W -Wall -Wextra -std=c++11 -I$(SRCDIR)/CPP
 
 CFLAGS ?= -O2
 CXXFLAGS ?= $(CFLAGS)
 
-7ZIP_CXX_SRC = CPP/7zip/Archive/Common/ParseProperties.cpp \
-               CPP/7zip/Archive/DeflateProps.cpp \
-               CPP/7zip/Common/CWrappers.cpp \
-               CPP/7zip/Common/FileStreams.cpp \
-               CPP/7zip/Common/InBuffer.cpp \
-               CPP/7zip/Common/OutBuffer.cpp \
-               CPP/7zip/Common/StreamUtils.cpp \
-               CPP/7zip/Compress/BitlDecoder.cpp \
-               CPP/7zip/Compress/DeflateDecoder.cpp \
-               CPP/7zip/Compress/DeflateEncoder.cpp \
-               CPP/7zip/Compress/LzOutWindow.cpp \
-               CPP/7zip/Compress/ZlibDecoder.cpp \
-               CPP/7zip/Compress/ZlibEncoder.cpp \
-               CPP/Common/MyString.cpp \
-               CPP/Common/StringConvert.cpp \
-               CPP/Common/StringToInt.cpp \
-               CPP/NotWindows/FileIO.cpp \
-               CPP/NotWindows/MyWindows.cpp \
-               CPP/Windows/PropVariant.cpp \
-               deflate7z.cpp
-7ZIP_CXX_OBJ = $(7ZIP_CXX_SRC:.cpp=.o)
-7ZIP_C_SRC =   C/Alloc.c \
-               C/HuffEnc.c \
-               C/LzFind.c \
-               C/Sort.c
-7ZIP_C_OBJ =   $(7ZIP_C_SRC:.c=.o)
-
-%.o: %.cpp
-	$(CXX) -c $(SRC_CXXFLAGS) $(CXXFLAGS) -o $@ $<
-
-%.o: %.c
+OBJDIR := objs
+BLDDIR := $(OBJDIR)/7zip
+MKDIRS := \
+	$(BLDDIR)/CPP/7zip/Archive/Common \
+	$(BLDDIR)/CPP/7zip/Common \
+	$(BLDDIR)/CPP/7zip/Compress \
+	$(BLDDIR)/CPP/Common \
+	$(BLDDIR)/CPP/NotWindows \
+	$(BLDDIR)/CPP/Windows \
+	$(BLDDIR)/C
+
+7ZIP_ARCHIVE_DIR := $(SRCDIR)/CPP/7zip/Archive
+7ZIP_COMMON_DIR := $(SRCDIR)/CPP/7zip/Common
+7ZIP_COMPRESS_DIR := $(SRCDIR)/CPP/7zip/Compress
```

Comment:
I'm not really sure what these variables are doing to make the makefile easier to read or edit.  Is there a reason why they're here for these three, but not i.e. for NotWindows?

-[Unknown]

Reply:
The only reason I added them was to avoid putting the same long path in multiple times, in the case for `CPP/Common`, `CPP/NotWindows` and `CPP/Windows` the paths were shorter so I thought it was unneeded. I guess I could add the others too if needed?

Reply:
Well, I suppose it's fine, I just prefer consistency.

-[Unknown]